create_layout.py

```diff
@@ -88,48 +88,118 @@
 
 """
 import os
+import itertools
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
-  """Create generic material rules (3 variants)
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a artifacts structure -- 'before' and 'after' -- 
+    return a tuple specifying which artifact have been added, 
+    which have been removed, which have been modified, 
+    and which have remained unchanged. 
+    Both these dictionaries have artifact
+    names as the keys and their hashes as the values.'''
 
-  * MATCH available materials with products from previous step (links must be an
+  before_set = set(before_dict.keys())
+  after_set = set(after_dict.keys())
+
+  removed_artifacts = before_set.difference(after_set)
+  added_artifacts = after_set.difference(before_set)
+
+  unchanged_artifacts = set()
+  modified_artifacts = set()
+  for key in before_set.intersection(after_set):
+    if before_dict[key] == after_dict[key]:
+      unchanged_artifacts.add(key)
+    else:
+      modified_artifacts.add(key)
+
+  # Returning the snapshot of the new artifacts system
+  return (unchanged_artifacts, modified_artifacts, added_artifacts,
+      removed_artifacts)
+
+def create_material_rules(previous_link, current_link):
+  """Create generic material rules
+
+  - MATCH available materials with products from previous step (links must be an
   ordered list) and
-  * ALLOW available materials if it is the first step in the
+  - ALLOW available materials if it is the first step in the
   list
-  Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
+  - DELETE removed materials
+
+  Args:
+    previous_link: a link of previous step,
+        including previous step's materials and prodcuts
+    current link: a link of current step,
+        including current step's materials and prodcuts
+
+  Returns:
+    a list of material rules 
+  """
 
   expected_materials = []
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = set(itertools.chain(unchanged_artifacts,
```

Comment:
Can't this use `.union` or `.extend` (if list)?

Reply:
Yes. `.union` works. Thanks for pointing this out!